core/src/main/scala/aima/core/agent/Agent.scala

```diff
@@ -3,26 +3,31 @@ package aima.core.agent
 /**
   * @author Shawn Garner
   */
-trait AgentProgram[Action, Percept] {
-  def actuators: List[Actuator[Action, Percept]]
-  def sensors: List[Sensor[Action, Percept]]
-  def agent: Agent[Action, Percept]
+trait Agent[ENVIRONMENT, PERCEPT, ACTION] {
+  def actuators: List[Actuator[ENVIRONMENT, ACTION]]
+  def sensors: List[Sensor[ENVIRONMENT, PERCEPT]]
+  def agentProgram: AgentProgram[PERCEPT, ACTION]
 
-  def run(environment: Environment[Action, Percept]): Environment[Action, Percept] = {
-    val actions = for {
-      sensor <- sensors
-    } yield agent.agentFunction(sensor.perceive(environment))
+  def run(e: ENVIRONMENT): (ENVIRONMENT, RunDetail[PERCEPT, ACTION]) = {
+    val percepts: List[PERCEPT] = sensors.flatMap(_.perceive(e))
+    val actions: List[ACTION]   = percepts.map(agentProgram.agentFunction)
 
-    actions.foldLeft(environment) { (env, action) =>
+    val newEnvironment = actions.foldLeft(e) { (env, action) =>
       actuators.foldLeft(env) { (env2, actuator) =>
-        actuator.act(action, env2)
+        actuator.act(action, env2).getOrElse(env2)
       }
     }
+    (newEnvironment, RunDetail(percepts, actions))
   }
 }
 
-trait Agent[Action, Percept] {
-  type AgentFunction = Percept => Action
+case class RunDetail[PERCEPT, ACTION](
```

Comment:
@benderpremier could we call this AgentRunSummary?  Detail makes it seem like we added more information.  Additionally I do wonder if we should have some algebra to for Summary which also includes a date time of the PERCEPT or the ACTION so you could order them somehow.

Reply:
Yeah I am happy to rename this to `AgentRunSummary`. I am not a native speaker so naming stuff correctly is never easy :) . 

Isn't the order of the percepts and actions already guaranteed by the elements of the list being processed in order ?

Reply:
@benderpremier They are ordered the way it's currently implemented however I could see different implementations where they might not be.  Also it's always nice to know when the agent perceived something and when it decided to perform an action.